src/core/lib/surface/server.cc

```diff
@@ -370,15 +371,25 @@ class Server::AllocatingRequestMatcherBatch
 
   void MatchOrQueue(size_t /*start_request_queue_index*/,
                     CallData* calld) override {
-    BatchCallAllocation call_info = allocator_();
-    GPR_ASSERT(server()->ValidateServerRequest(
-                   cq(), static_cast<void*>(call_info.tag), nullptr, nullptr) ==
-               GRPC_CALL_OK);
-    RequestedCall* rc = new RequestedCall(
-        static_cast<void*>(call_info.tag), call_info.cq, call_info.call,
-        call_info.initial_metadata, call_info.details);
-    calld->SetState(CallData::CallState::ACTIVATED);
-    calld->Publish(cq_idx(), rc);
+    server()->shutdown_blocking_refs_.fetch_add(1, std::memory_order_acq_rel);
+    if (!server()->shutdown_flag_.load(std::memory_order_acquire)) {
```

Comment:
There are two atomics here. I'm still trying to draw all happen before/after links.

Is there a way to somehow join them into one variable? Otherwise, it would be somewhat difficult to reason about this.

Reply:
Yes, let's do this. I'll spend the time to resolve this as it makes the code too complex otherwise.

Reply:
Done, using the evens-and-odds technique with the LSB representing whether shutdown has not been called and the other bits representing the actual number of in-progress requests.